module/os/linux/zfs/abd.c

```diff
@@ -802,15 +978,99 @@ abd_alloc_for_io(size_t size, boolean_t is_metadata)
 	return (abd_alloc(size, is_metadata));
 }
 
+/*
+ * Create an ABD that will be the head of a list of ABD's. This is used
+ * to "chain" scatter/gather lists together when constructing aggregated
+ * IO's. To free this abd, abd_put() must be called.
+ */
+abd_t *
+abd_alloc_multi(void)
+{
+	abd_t *abd;
+
+	abd = abd_alloc_struct();
+	abd->abd_flags = ABD_FLAG_MULTI_LIST;
+	abd->abd_size = 0;
+	abd->abd_parent = NULL;
+	list_create(&ABD_MULTI(abd).abd_chain,
+	    sizeof (abd_t), offsetof(abd_t, multi_link));
+	zfs_refcount_create(&abd->abd_children);
+	return (abd);
+}
+
+/*
+ * Add a child ABD to a chained list of ABD's.
+ */
+void
+abd_add_child(abd_t *pabd, abd_t *cabd, boolean_t multi_mem_manage)
+{
+	ASSERT(abd_is_multi(pabd));
+	abd_t *child_abd = NULL;
+
+	mutex_enter(&cabd->abd_mtx);
+	if (list_link_active(&cabd->multi_link)) {
```

Comment:
If `multi_mem_manage` is TRUE, and `list_link_active()` is also TRUE, I don't think we implement the caller's request to free the passed-in ABD when the multi-abd is freed.  I'm not sure how we could implement it.

Also, if the cabd's link is active, and ABD_FLAG_MULTI_FREE is already set on it, I'm not sure how that would work, depending on which order the two multi-abd's are freed.

In general, the `multi_mem_manage=TRUE` case does not seem like it can be very general purpose.  I think we need to at least ASSERT that you can't combine `multi_mem_manage=TRUE` with having the child be part of multiple multi-abd's.  It seems a bit fragile, so maybe we should explain the `multi_mem_manage=TRUE` use case a bit more, and the restrictions on it.

Or, I wonder if we should get rid of `multi_mem_manage=TRUE` (for callers outside the abd code), and make them free all the child abd's themself.  E.g. in `vdev_queue_agg_io_done()`, like we're already doing for the "write gap" abd's.

Reply:
I agree that an ASSERT is absolutely necessary here. In general, if someone passes TRUE here they are explicitly saying they expect the multilist abd to take care of the memory management on the call to abd_free(). So by placing the ASSERT it will definitely signal an issue when someone tries to use the same abd in multiple multilist abd’s abs still expects all the memory management to be taken care of in the call to abd_free() on the multilist abd. I will also explain this more in the comments as well.

I kinda like flagging if the multilist abd should free the underlying abd’s. It allows for possibly easier use of the multilist abd by just being able to call abd_free() if it is known that freeing all the underlying abd’s can be handled there.

Reply:
I still kept the parameter for "free_on_free" here. I did however add an ASSERT with a comment explaining "free_on_free" must be B_FALSE if an ABD is already part of another Gang ABD.